src/Core/Context/LegacyControllerContextBuilder.php

```diff
@@ -0,0 +1,224 @@
+<?php
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Core\Context;
+
+use Doctrine\ORM\NoResultException;
+use PrestaShop\PrestaShop\Adapter\ContextStateManager;
+use PrestaShop\PrestaShop\Adapter\Feature\MultistoreFeature;
+use PrestaShop\PrestaShop\Core\Domain\Shop\ValueObject\ShopConstraint;
+use PrestaShop\PrestaShop\Core\Util\Inflector;
+use PrestaShopBundle\Entity\Repository\TabRepository;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Contracts\Translation\TranslatorInterface;
+use Tools;
+
+class LegacyControllerContextBuilder implements LegacyContextBuilderInterface
+{
+    public function __construct(
+        private readonly EmployeeContext $employeeContext,
+        private readonly ShopContext $shopContext,
+        private readonly ContextStateManager $contextStateManager,
+        private readonly RequestStack $requestStack,
+        private readonly array $controllersLockedToAllShopContext,
+        private readonly TabRepository $tabRepository,
+        private readonly TranslatorInterface $translator,
+        private readonly MultistoreFeature $multistoreFeature,
+        private readonly ContainerInterface $container,
+        private readonly string $coreDir,
+    ) {
+    }
+
+    public function buildLegacyContext(): void
+    {
+        $request = $this->requestStack->getMainRequest();
+
+        $controllerName = $this->getControllerName($request);
+        $multiShopContext = $this->getMultiShopContext($controllerName);
+        $id = $this->getTabId($controllerName);
+        $token = null;
+        if ($this->employeeContext->getEmployee()) {
+            $token = Tools::getAdminToken($controllerName . $id . $this->employeeContext->getEmployee()->getId());
```

Comment:
I don't remember the other use cases exactly, but I believe there are cases in the token generation where the employee ID is not available (or sometimes the tab ID) but the token can still be generated, it's just that the employeeId part of the generated key remains empty no?

Reply:
Ah found it https://github.com/PrestaShop/PrestaShop/blob/4d49949dbf350e79f4253191707a2283ad54f328/src/PrestaShopBundle/Twig/Component/HeadTag.php#L172

By the way, I believe we have multiple places where we handle this legacy token generation, it might be worth it having a single dedicated class/service that can do that to avoid duplicate code and to factorize it Not necessarily in this PR but then we must prepare an issue to do this later

Which makes me think we may have multiple todos to handle after the main issues will be done.

Reply:
Are you thinking of the _PS_ADMIN_DIR_ constant? :smiling_face_with_tear: